R/data.table.R

```diff
@@ -2815,3 +2764,136 @@ gforce <- function(env, jsub, o, f, l, rows) .Call(Cgforce, env, jsub, o, f, l,
+      # re-direct all non-matching mode cases to base R, as data.table's binary
+      # search based join is strict in types. #957 and #961.
+      return(NULL)
+    }
+    if (operator == "==") { ## addional requirements for equal operator and notjoin
+      # RHS is of length=1 or n
+      if (any_na(as_list(RHS))) {
+        notjoin = FALSE
+        RHS = RHS[0L]
+      } else if (notjoin) {
+        RHS = c(RHS, if (is.double(RHS)) c(NA, NaN) else NA)
+      }
+    }
+    ## if it passed until here, fast subset can be done for this stub
+    i <- c(i, setNames(list(RHS), col))
+    ## loop continues with remainingIsub
+  } 
+  if (length(i) == 0) stop("Internal error in .isFastSubsettable. Please report to data.table developers")
+  ## convert i to data.table with all combinations in rows. Care is needed with names as we do 
+  ## it with 'do.call' and this would cause problems if colNames were 'sorted' or 'unique'
+  ## as these two would be interpreted as args for CJ
+  colNames <- names(i)
+  names(i) <- NULL 
+  i$sorted <- TRUE
+  i$unique <- TRUE
+  i <- do.call(CJ, i)
+  setnames(i, colNames)
+  idx <- NULL 
+  ## check whether key fits the columns in i.
+  ## order of key columns makes no difference, as long as they are all upfront in the key, I believe.
+  if (all(names(i) %chin% head(key(x), length(i)))){
+      if (verbose) {cat("Optimized subsetting with key '", paste0( head(key(x), length(i)), collapse = ", "),"'\n",sep="");flush.console()}
+      idx <- integer(0) ## integer(0) is not NULL! Indicates that is is ordered correctly.
+      idxCols <- head(key(x), length(i)) ## in correct order!
+  } 
+  if (is.null(idx)){
+    ## check whether an exising index can be used
+    ## An index can be used if the first elements correspond to the columns in i (similar to the key above)
+    ## Previously, the head of an index was not used because of reordering within groups of i.
+    ## However, I believe this is of no concern as the original order is restored after the bmerge anyways.
+    ## if multiple indices are possible, choose one with integer(0), so no reordering is necessary
+    candidates <- indices(x, vectors = TRUE)
+    idx <- NULL
+    for (cand in candidates){
+      if (all(names(i) %chin% head(cand, length(i)))){
+        idx <- attr(attr(x, "index"), paste0("__", cand, collapse = ""))
+        finalCand <- cand
+        if (length(idx) == 0) break ## the index is already optimal since it doesn't require sorting
+      }
+    }
+    if (!is.null(idx)){
+      if (verbose) {cat("Optimized subsetting with index '", paste0("__", finalCand, collapse = ""),"'\n",sep="");flush.console()}
+      idxCols <- head(finalCand, length(i)) ## in correct order!
+    }
+  } 
+  if (is.null(idx)){
+    ## if nothing else helped, auto create a new index that can be used
+    if (!getOption("datatable.auto.index")) return(NULL) 
```

Comment:
just for information... on the other hand this option allows to use indexes only when they were created manually by calling `setindex` (or created before setting up this option).